infrastructure/src/main/java/org/corfudb/infrastructure/LogReplicationServer.java

```diff
@@ -94,43 +86,59 @@ public void shutdown() {
 
     /* ************ Server Handlers ************ */
 
-    @ServerHandler(type = CorfuMsgType.LOG_REPLICATION_ENTRY)
-    private void handleLogReplicationEntry(CorfuPayloadMsg<LogReplicationEntry> msg, ChannelHandlerContext ctx, IServerRouter r) {
+    /**
+     * Given a log-entry request message, send back the log-data entries.
+     *
+     * @param request leadership query
+     * @param ctx     enables a {@link ChannelHandler} to interact with its
+     *                {@link ChannelPipeline} and other handlers
+     * @param router  router used for sending back the response
+     */
+    @RequestHandler(type = PayloadCase.LR_ENTRY_REQUEST)
+    private void handleLrEntryRequest(@Nonnull RequestMsg request,
+                                      @Nonnull ChannelHandlerContext ctx,
+                                      @Nonnull IServerRouter router) {
         log.trace("Log Replication Entry received by Server.");
 
-        if (isLeader(msg, r)) {
+        if (isLeader(request, ctx, router)) {
             // Forward the received message to the Sink Manager for apply
-            LogReplicationEntry ack = sinkManager.receive(msg.getPayload());
+            LogReplicationEntryMsg ack =
+                    sinkManager.receive(request.getPayload().getLrEntryRequest());
 
             if (ack != null) {
-                long ts = ack.getMetadata().getMessageMetadataType().equals(MessageType.LOG_ENTRY_REPLICATED) ?
+                long ts = ack.getMetadata().getEntryType().equals(LogReplicationEntryType.LOG_ENTRY_REPLICATED) ?
                         ack.getMetadata().getTimestamp() : ack.getMetadata().getSnapshotTimestamp();
                 log.info("Sending ACK {} on {} to Client ", ack.getMetadata(), ts);
-                r.sendResponse(msg, CorfuMsgType.LOG_REPLICATION_ENTRY.payloadMsg(ack));
+                ResponsePayloadMsg payload = getLogEntryResponse(request);
+                ResponseMsg response = getResponseMsg(request.getHeader(), payload);
+                router.sendResponse(response, ctx);
```

Comment:
Mostly due to hierarchy. The implementation is overriding `IServerRouter::sendResponse(ResponseMsg response, ChannelHandlerContext ctx)` and this is done in `LogReplicationServerRouter`. Two other solutions that I see:

1. Define a separate function `IServerRouter::sendResponse(ResponseMsg response)` that does not include the context. The problem with this approach is that any one implementation would be implementing one of those two methods. So it becomes unclear which method is implemented and which one is not.
 
2. Make `LogReplicationServerRouter` implement a new interface that is not `IServerRouter`. This is an ideal solution, but I wanted to make the patch non-intrusive. I you prefer, I can make that change in this patch.

Reply:
Yes, I think solution 2 would be nice, otherwise it is a bit misleading... as you prefer, if you think you can add it in this patch great!

Reply:
Turns out when creating a new `IServerRouter` interface to excludes Netty related constructs, we also need to change `AbstractServer`, `RequestHandlerMethods` as well as the `ServerContext`. Seems like we first need to decouple those four classes before we can create a new interface. 

Since this will not be a mechanical change, I suggest I address this in a separate PR.